File: Algorithms Advanced/others/GraphAlgorithms/Python/Dijkstra.py
import heapq

def dijkstra(start, adj, dist):
    pq = [(0, start)]
    dist[start] = 0

    while pq:
        d, node = heapq.heappop(pq)

        if d > dist[node]:
            continue

        for neighbor, weight in adj[node]:
            if dist[node] + weight < dist[neighbor]:
                dist[neighbor] = dist[node] + weight
                heapq.heappush(pq, (dist[neighbor], neighbor))

# Example
nodes = 5
adj = {0: [(1, 2), (4, 1)], 1: [(2, 3)], 4: [(3, 2)], 3: [(2, 1)]}
dist = [float('inf')] * nodes
dijkstra(0, adj, dist)

for i in range(nodes):
    print(f"Distance to {i}: {dist[i]}")

File: Algorithms Advanced/others/GraphAlgorithms/C++/Dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

void dijkstra(int start, vector<vector<pair<int, int>>> &adj, vector<int> &dist) 
{
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    pq.push({0, start});
    dist[start] = 0;

    while (!pq.empty()) 
    {
        int d = pq.top().first;
        int node = pq.top().second;
        pq.pop();

        if (d > dist[node]) continue;

        for (auto neighbor : adj[node]) 
        {
            int next = neighbor.first;
            int weight = neighbor.second;

            if (dist[node] + weight < dist[next]) 
            {
                dist[next] = dist[node] + weight;
                pq.push({dist[next], next});
            }
        }
    }
}

int main() 
{
    int nodes = 5;
    vector<vector<pair<int, int>>> adj(nodes);
    adj[0].push_back({1, 2});
    adj[0].push_back({4, 1});
    adj[1].push_back({2, 3});
    adj[4].push_back({3, 2});
    adj[3].push_back({2, 1});

    vector<int> dist(nodes, INT_MAX);
    dijkstra(0, adj, dist);

    for (int i = 0; i < nodes; i++) 
    {
        cout << "Distance to " << i << ": " << dist[i] << endl;
    }
    return 0;
}
